dlib/dnn/trainer.h

```diff
@@ -1057,34 +1076,67 @@ namespace dlib
             if (!std::ifstream(newest_syncfile(), std::ios::binary))
                 return false;
 
-            for (auto x : previous_loss_values)
+            // if we haven't seen much data yet then just say false.
+            if (gradient_updates_since_last_sync < 30)
+                return false;
+
+            // how long a loss history do we actually need to maintain?
+            const size_t max_previous_loss_values_to_consider = 2 * gradient_updates_since_last_sync;
+            const auto total_previous_loss_values = [this]()
+            {
+                return additional_previous_loss_values_to_keep_until_disk_sync.size()
+                     + previous_loss_values.size();
+            };
+            const auto can_drop_an_additional_previous_loss_value = [&]() {
+                return !additional_previous_loss_values_to_keep_until_disk_sync.empty()
+                    && total_previous_loss_values() > max_previous_loss_values_to_consider;
+            };
+
+            while (can_drop_an_additional_previous_loss_value())
+                additional_previous_loss_values_to_keep_until_disk_sync.pop_front();
+
+            // collect all previous loss values we want to consider
```

Comment:
This is pretty much the beef of the trainer code change introduced in the PR summary.

Reply:
(Certainly not saying these are the exact changes that we need, but at a bare minimum I want to start a discussion here.)

Reply:
If I understand, the issue this is trying to address is that when we lower the learning rate, the code also kicks a bunch of values out of `previous_loss_values`, which then forces `loss_increased_since_last_disk_sync()` to return `false` for a while no matter what. I agree that that's not great and `loss_increased_since_last_disk_sync()` shouldn't ever be inhibited.

I think we can make this code a lot simpler though.  Recall that the reason lowering the learning rate kicked values out of `previous_loss_values` was just to prevent the learning rate from being immediately lowered again based on the same analysis of `previous_loss_values`. That was a crude way of doing it.  So rather than adding `additional_previous_loss_values_to_keep_until_disk_sync` and all the attendant extra logic, how about when the loss stalls we don't kick any values out of `previous_loss_values` but instead just update a counter that is used to measure the number of steps since the last time we lowered the learning rate.  That counter can then be used to inhibit execution of the code at line 800, by changing it from this:
```
                if (gradient_check_budget > iter_without_progress_thresh && learning_rate_shrink != 1)
 ```
to something like this:
```
                if (gradient_check_budget > iter_without_progress_thresh && learning_rate_shrink != 1 && steps_since_last_learning_rate_shrink > iter_without_progress_thresh)
 ```

This also means the code below there that does this:
```
                            // Empty out some of the previous loss values so that steps_without_progress 
                            // will decrease below iter_without_progress_thresh.  
                            for (unsigned long cnt = 0; cnt < previous_loss_values_dump_amount+iter_without_progress_thresh/10 && previous_loss_values.size() > 0; ++cnt)
                                previous_loss_values.pop_front();
```
would be replaced with something like 
```
steps_since_last_learning_rate_shrink -= previous_loss_values_dump_amount+iter_without_progress_thresh/10;
```

Reply:
> If I understand, the issue this is trying to address is that when we lower the learning rate, the code also kicks a bunch of values out of previous_loss_values, which then forces loss_increased_since_last_disk_sync() to return false for a while no matter what.

Correct.

> I think we can make this code a lot simpler though.

Thanks for the suggestion – will try it.

Reply:
Hmm. The current `master` code limits the length of `previous_loss_values` to no more than `iter_without_progress_thresh`. Now let's say `iter_without_progress_thresh` is relatively low, when compared to the disk sync interval. The PR, as it stands right now, addresses this issue also (and not only the explicit kick that may happen when the learning rate is shrunk).

Reply:
I’m not hugely worried about that case. The user is supposed to set the threshold such that it’s long enough to tell what the loss is doing. Either going up or down. 

Although, I would not be adverse to there being two separate loss histories. One used for the check if the loss is going up and another for the check about going down. There isn’t any reason why the one used for the going up check can’t be the entire length since last sync. And that would probably be better, as you point out. 

I just want the code to be simple. Using two totally separate loss histories seems like it would allow a simple implementation.

Reply:
> I’m not hugely worried about that case. The user is supposed to set the threshold such that it’s long enough to tell what the loss is doing. Either going up or down.

Well I already shot myself in the foot, and I'm certainly worried about doing it again. In particular when one thing (disk sync interval) is specified in wall-clock time, and the other (iterations without progress) in steps.

> I just want the code to be simple. Using two totally separate loss histories seems like it would allow a simple implementation.

Yeah, I too want it to be simple. Tried to avoid saving the values two times, but will happily backtrack.

Reply:
Not a big deal to save extra stuff in the sync file. It happens rarely and the file never goes anywhere or gets stored long term. So don’t worry about putting redundant stuff in there. Whatever makes the code the simplest is probably the best. :)

Reply:
Simplified the code as discussed, and tested the result. Seems to be ok; pushed.